### PR TITLE
[feature] Added raw id filter #248

### LINKED ISSUE #248: [feature] Add custom raw ID filter
Move the logic of the filter added for subnet division rule and payments in the subscriptions module in this module to avoid duplication and ease reusability.

### PR DESCRIPTION
**Pull request by niteshsinha17**
This filter can be used in other modules to add a raw id filter
#### Checklist

- [x] I have read the [contributing guidelines](http://openwisp.io/docs/developer/contributing.html#how-to-commit-your-changes-properly).
- [x] I have manually tested the proposed changes.
- [ ] I have written new test cases to avoid regressions. (if necessary)
- [ ] I have updated the documentation. (e.g. README.rst)
- [ ] I have added [change!] to commit title to indicate a backward incompatible change. (if required)
- [ ] I have checked the links added / modified in the documentation.

closes #248

### CODE REVIEW

**Review by nemesisdesign — COMMENTED** [review]
@niteshsinha17 could you write a test that covers these 2 lines please?
https://coveralls.io/builds/42860654/source?filename=openwisp_utils%2Fadmin_theme%2Ffilter.py#L16
So we can avoid the build failure for coverage decreasing, albeit is small but it's nice to cover all new lines.

File: openwisp_utils/admin_theme/static/admin/js/ow-filter.js [hunk 1]
```diff
@@ -164,6 +165,29 @@ function setArrowButtonVisibility() {
   }
 }
 
+function initInputFilterHandler() {
```
Comment (pandafy) [inline openwisp_utils/admin_theme/static/admin/js/ow-filter.js:168]: This is great!

**Review by pandafy — CHANGES_REQUESTED** [review]
Good work @niteshsinha17 👍🏼 

I noticed some minor issues with code formatting. We can merge this one those are addressed.

File: openwisp_utils/admin_theme/static/admin/js/ow-filter.js [hunk 2]
```diff
@@ -239,6 +266,16 @@ function filterHandlers() {
...
+    inputFilters.forEach(function (filter){
+      var href = filter.getAttribute('href');
+      var [key, val] = href.substring(1).split('=');
+      if(val.length === 0 && key in qs){
+        
+        delete qs[key];
+      }else if(val.length !== 0) {
+        qs[key] = val;
+      }
+    });
```
[the remark this answers is not in the document]
Comment (nemesisdesign) [reply openwisp_utils/admin_theme/static/admin/js/ow-filter.js:278]: don't we have style checks for JS code?

[the remark this answers is not in the document]
Comment (niteshsinha17) [reply openwisp_utils/admin_theme/static/admin/js/ow-filter.js:278]: It checks on basic formatting.

File: openwisp_utils/admin_theme/static/admin/js/ow-filter.js [hunk 3]
```diff
@@ -189,7 +213,10 @@ function filterHandlers() {
     const selectedOptions = document.querySelectorAll(
       '.filter-options .selected'
     );
-    // Create params map which knows about the last applied filters
+    const inputFilters = document.querySelectorAll(
+      '.ow-input-filter .filter-options a'
+    );
+    // Create params 1map which knows about the last applied filters
```
[the remark this answers is not in the document]
Comment (niteshsinha17) [reply openwisp_utils/admin_theme/static/admin/js/ow-filter.js]: Thanks

File: openwisp_utils/admin_theme/templates/admin/input_filter.html [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+<div class="ow-filter ow-input-filter">
+  <div class="filter-title" title="{{ choice.display }}">
+    {% with choices.0 as all_choice %}
+    <h3>{% blocktrans with filter_title=title %} By {{ filter_title }} {% endblocktrans %}</h3>
+    <form method="GET" action="">
+      {% for key, value in all_choice.query_parts %}
+      <input type="hidden" name="{{ key }}" value="{{ value }}" />
+      {% endfor %}
+      <input type="text" class="ow-input-filter-field" title="{{ spec.value | default_if_none:'' }}"value="{{ spec.value | default_if_none:'' }}"
```
Comment (nemesisdesign) [inline openwisp_utils/admin_theme/templates/admin/input_filter.html:11]: a space is missing here before `value`.

I recommend breaking the attributes over new lines, eg:

```html
<input type="text"
       class="...."
       title="..."
       value="..."
/>
```

File: openwisp_utils/admin_theme/templates/admin/input_filter.html [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        name="{{ spec.parameter_name }}" />
+      {% if not all_choice.selected %}
+      <a class='field-clear' title="{% trans 'Clear' %}" href="{{ all_choice.query_string }}">&#x2716; </a>
```
Comment (nemesisdesign) [inline openwisp_utils/admin_theme/templates/admin/input_filter.html:14]: same here, I would break this line

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Thanks @niteshsinha17 @pandafy!

See my comments below for some minor code improvements.

Definitely let's add a mention of this in the README, like at the end of this section: https://github.com/openwisp/openwisp-utils#admin-utilities, a screenshot in the media branch would be highly appreciated: https://github.com/openwisp/openwisp-utils/tree/media/docs.

File: tests/test_project/admin.py [hunk 6]
```diff
@@ -57,12 +58,21 @@ class ProjectAdmin(UUIDAdmin, ReceiveUrlAdmin):
     receive_url_name = 'receive_project'
 
 
+class ShelfFilter(InputFilter):
+    parameter_name = 'shelf'
+    title = _('Shelf')
+
+    def queryset(self, request, queryset):
+        if self.value() is not None:
+            return queryset.filter(name__icontains=self.value())
```
[the remark this answers is not in the document]
Comment (niteshsinha17) [reply tests/test_project/admin.py:67]: Yes need to be overridden,

File: openwisp_utils/admin_theme/templates/admin/input_filter.html [hunk 7]
```diff
@@ -0,0 +1,32 @@
+{% load i18n %}
+
+<div class="ow-filter ow-input-filter">
+  <div class="filter-title" title="{{ choice.display }}">
+  {% with choices.0 as all_choice %}
+    <h3>{% blocktrans with filter_title=title %} By {{ filter_title }} {% endblocktrans %}</h3>
```
Comment (pandafy) [inline openwisp_utils/admin_theme/templates/admin/input_filter.html:6]:
```suggestion
    <h3>{% blocktrans with filter_title=title %}By {{ filter_title }}{% endblocktrans %}</h3>
```

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
I did some testing and looked at the code, it's good to merge except a few minor whitespace issues, see below.

**Review by nemesisdesign — APPROVED** [review]

**Pull request merged**